File: scala/27.scala
// JAVA_OPTS="-Xmx1024M" time scala -deprecation 27.scala
// -59231
// 10.47user 0.88system 0:14.85elapsed 76%CPU (0avgtext+0avgdata 0maxresident)k
// 0inputs+0outputs (0major+82318minor)pagefaults 0swaps

import scala.io.Source

def f(n: Int, a: Int, b: Int) = (n * n) + (a * n) + b

def count(ps: Set[Int], a: Int, b: Int): Int = {
  (0 until b).takeWhile(ps contains f(_, a, b)).size
}

def primes = Set() ++ Source.fromFile("../data/primes.txt")
                            .getLines
                            .map(_.trim.toInt)

val ps = primes.filter(_ < 1000)
val m = Map() ++ {
  for (
    a <- -1000 to 1000;
    b <- ps ++ ps.map(-1 * _);
    c =  count(ps, a, b)
  ) yield c -> (a, b)
}

println {
  val tuple = m(m.keys.max)
  tuple._1 * tuple._2
}

File: ruby/27.rb
# time ruby 27.rb
# -59231
# 11.83user 1.42system 0:13.79elapsed 96%CPU (0avgtext+0avgdata 0maxresident)k
# 0inputs+0outputs (0major+2449minor)pagefaults 0swaps

require 'set'

# A simple quadratic
def f(n, a, b)
  n*n + a*n + b
end

# For how many values is a quadratic prime?
def count(ps, a, b)
  (0..b).take_while { |n| ps.include? f(n, a, b) }.count
end

# A list of the primes under 2 million will be fine
def primes
  file = File.new("../data/primes.txt")
  set = Set.new
  while (p = file.gets)
    set.add(p.to_i)
  end
  file.close
  set
end


b_max   = 1000
ps      = primes
bs      = ps.select { |b| b <= b_max }
bs      = bs + bs.map { |b| -b }
a, b, c = 0, 0, 0

bs.each do |b_i|
  (-b_max..b_max).each do |a_i|
    c_i = count(ps, a_i, b_i)
    if c_i > c
      a, b, c = a_i, b_i, c_i
    end
  end
end

puts a*b
